quil-rs/src/program/mod.rs

```diff
@@ -348,6 +360,90 @@ impl Program {
         Ok(expanded_program)
     }
 
+    /// Resolve [`LabelPlaceholder`]s and [`QubitPlaceholder`]s within the program such that the resolved values
+    /// will remain unique to that placeholder within the scope of the program.
+    ///
+    /// If you provide `label_resolver` and/or `qubit_resolver`, those will be used to resolve those values respectively.
+    /// If your placeholder returns `None` for a particular placeholder, it will not be replaced but will be left as a placeholder.
+    ///
+    /// If you do not provide a resolver for a placeholder, a default resolver will be used which will generate a unique value
+    /// for that placeholder within the scope of the program using an auto-incrementing value (for qubit) or suffix (for label)
+    /// while ensuring that unique value is not already in use within the program.
+    #[allow(clippy::type_complexity)]
+    pub fn resolve_placeholders(
+        &mut self,
+        label_resolver: Option<Box<dyn Fn(&LabelPlaceholder) -> Option<String>>>,
+        qubit_resolver: Option<Box<dyn Fn(&QubitPlaceholder) -> Option<u64>>>,
+    ) {
+        let qubit_resolver = qubit_resolver.unwrap_or_else(|| {
+            let mut qubits_used: HashSet<u64> = HashSet::new();
+            let mut qubit_placeholders: IndexSet<QubitPlaceholder> = IndexSet::new();
+
+            // Stable iteration order makes placeholder resolution deterministic
+            let sorted_qubits = self.get_used_qubits().iter().collect::<BTreeSet<_>>();
+
+            for qubit in sorted_qubits {
+                match qubit {
+                    Qubit::Fixed(index) => {
+                        qubits_used.insert(*index);
+                    }
+                    Qubit::Placeholder(placeholder) => {
+                        qubit_placeholders.insert(placeholder.clone());
+                    }
+                    _ => {}
+                }
+            }
+
+            let mut qubit_iterator = (0u64..).filter(|index| !qubits_used.contains(index));
+            let qubit_resolutions: HashMap<QubitPlaceholder, u64> = qubit_placeholders
+                .into_iter()
+                .map(|p| (p, qubit_iterator.next().unwrap()))
+                .collect();
```

Comment:
```suggestion
            let qubit_resolutions: HashMap<QubitPlaceholder, u64> = qubit_placeholders
                .into_iter()
                .zip(0u64..)
                .collect();
```

Reply:
Not equivalent, because of the lack of filter. Could still zip the qubit iterator though